### PR TITLE
indexdata: read posting list iff all ng exist

### PR DESCRIPTION
**Pull request by stefanhengl**
The purpose of this PR is to reduce disk IO in case we skip a shard because of missing ngrams.

To achieve this, we first check whether ALL ngrams exist in the shard before loading the posting lists to determine their frequency. This means we have to loop twice over the ngrams for the benefit of not loading any posting list in case the shard would have been skipped anyways. 

Test plan:
- This is a refactor, so relying on CI

### CODE REVIEW

File: indexdata.go [hunk 1]
```diff
@@ -387,32 +407,30 @@ func (d *indexData) iterateNgrams(query *query.Substring) (*ngramIterationResult
...
-			}, nil
-		}
+			},
+		}, nil
+	}
 
-		frequencies = append(frequencies, freq)
+	frequencies := make([]uint32, 0, len(ngramOffs))
+	for _, ngramIndex := range ngramIndexes {
+		frequencies = append(frequencies, d.ngramIndexFrequency(ngramIndex, query.FileName))
```
Comment (keegancsmith) [inline indexdata.go]: you might as well make ngramIndexFrequency take in the ngramIndexes slice and avoid the loop here.

**Review by keegancsmith — COMMENTED** [review]
this seems pretty tricky. I wonder if we could design this slightly differently. These are the concerns:

1. btree code caring about caseSensitive. I understand why we need to do this, but it kinda feels wrong. Would this code be better if we only did this sort of optimization for case sensitive search?
2. allocations. I'm unsure on the impact of the GC since this does a bunch more allocations. It's not so much size I am worried about, but number of objects. In particular [][]int will have len() objects.
3. Is this worth it? I suppose if it halves IO for attribution search it is.

I guess the other solution was changing how we laid out data on disk, but that requires a reindex while this doesn't which makes it appealing. What are your feelings about this after working on it?


I'm mulling on this a bit. And spending allocations is likely the right tradeoff. If you rule out a shard (the common case) that saved IO is good. If we don't rule it out, then more than likely we will be doing many more allocations while actually searching it.

Then there is two bits of future work that could likely help here:
1. If this is good, change the shape of our data on disk so ngram is stored with posting list location
2. The stuff we do for IR introducing opportunity for reusing buffers/etc

**Comment by stefanhengl** [comment]
> Would this code be better if we only did this sort of optimization for case sensitive search?

Is attribution search case sensitive?

**Comment by keegancsmith** [comment]
> > Would this code be better if we only did this sort of optimization for case sensitive search?
> 
> Is attribution search case sensitive?

yes. right now it is (mainly for speed)

**Comment by stefanhengl** [comment]
> > > Would this code be better if we only did this sort of optimization for case sensitive search?
> > 
> > 
> > Is attribution search case sensitive?
> 
> yes. right now it is (mainly for speed)

Alright. Then let's focus on case sensitive first. Once we have other optimisations in place we can revisit.

File: indexdata.go [hunk 2]
```diff
@@ -412,7 +431,35 @@ func (d *indexData) iterateNgrams(query *query.Substring) (*ngramIterationResult
...
+		for _, ngramIndex := range ngramIndexes {
+			frequencies = append(frequencies, d.ngramIndexFrequency(ngramIndex, query.FileName))
+		}
+	} else {
+		for _, o := range ngramOffs {
+			var freq uint32
+			if query.CaseSensitive {
+				freq = d.ngramFrequency(o.ngram, query.FileName)
+				ngramLookups++
+			} else {
```
Comment (keegancsmith) [inline indexdata.go]: we can remove this code path?

**Review by keegancsmith — APPROVED** [review]
LGTM. How do you think we can validate if this makes a difference? experiments on production and then checking the raw search time? IE the same thing I am doing with the sorting change?

**Review by keegancsmith — APPROVED** [review]
LGTM. How do you think we can validate if this makes a difference? experiments on production and then checking the raw search time? IE the same thing I am doing with the sorting change?

**Comment by stefanhengl** [comment]
> LGTM. How do you think we can validate if this makes a difference? experiments on production and then checking the raw search time? IE the same thing I am doing with the sorting change?

I think so. Let's finish validating your changes first. I will wait with merging until then.

**Comment by keegancsmith** [comment]
@stefanhengl I merged all my PRs which horribly conflicted with this. So I went and resolved it for ya, please take a look at the changes to makes sure I didn't mess it up.

**Comment by keegancsmith** [comment]
I am gonna merge this now so we can lay the ground work to get this out into production tomorrow.

**Pull request merged**